common/filesystem.cpp

```diff
@@ -0,0 +1,515 @@
+  for (;;) {
+    auto read_result = ReadToBuffer(buffer);
+    if (!read_result.ok()) {
+      return std::move(read_result).error();
+    }
+    if (read_result->empty()) {
+      // EOF
+      break;
+    }
+    result.append(reinterpret_cast<const char*>(read_result->data()),
+                  read_result->size());
+  }
+
+  return result;
+}
+
+auto Internal::FileRefBase::WriteFromString(llvm::StringRef str)
+    -> ErrorOr<Success, FdError> {
+  auto bytes = llvm::ArrayRef<std::byte>(
+      reinterpret_cast<const std::byte*>(str.data()), str.size());
+  while (!bytes.empty()) {
+    auto write_result = WriteFromBuffer(bytes);
+    if (!write_result.ok()) {
+      return std::move(write_result).error();
+    }
+    bytes = *write_result;
+  }
+  return Success();
+}
+
+auto DirRef::ReadFileToString(std::filesystem::path path)
+    -> ErrorOr<std::string, PathError> {
+  CARBON_ASSIGN_OR_RETURN(ReadFile f, OpenReadOnly(path));
+  auto result = f.ReadToString();
+  if (result.ok()) {
+    return *std::move(result);
+  }
+  return PathError(result.error().number(),
+                   "Dir::ReadFileToString on '{0}' relative to '{1}'",
+                   std::move(path), dfd_);
+}
+
+auto DirRef::WriteFileFromString(std::filesystem::path path,
+                                 llvm::StringRef content,
+                                 CreationFlags creation_flags)
+    -> ErrorOr<Success, PathError> {
+  CARBON_ASSIGN_OR_RETURN(WriteFile f, OpenWriteOnly(path, creation_flags));
+  auto result = f.WriteFromString(content);
+  if (result.ok()) {
+    return Success();
+  }
+  return PathError(result.error().number(),
+                   "Dir::WriteFileFromString on '{0}' relative to '{1}'",
+                   std::move(path), dfd_);
+}
+
+auto DirRef::CreateDirectories(std::filesystem::path path,
+                               ModeType creation_mode)
+    -> ErrorOr<Dir, PathError> {
+  // Avoid having to handle an empty path by immediately rejecting it as
+  // invalid.
+  if (path.empty()) {
+    return PathError(EINVAL,
+                     "DirRef::CreateDirectories on '{0}' relative to '{1}'",
+                     std::move(path), dfd_);
+  }
+  // Try directly opening the directory and use that if successful. This is an
+  // important hot path case of users essentially doing an "open-always" form of
+  // creating multiple steps of directories.
+  auto open_result = OpenDir(path, OpenExisting);
+  if (open_result.ok()) {
+    return std::move(*open_result);
+  } else if (!open_result.error().no_entity()) {
+    return std::move(open_result).error();
+  }
+
+  // Walk from the full path towards this directory (or the root) to find the
+  // first existing directory. This is faster than walking down as no file
+  // descriptors have to be allocated for any intervening directories, etc. We
+  // keep the path components that are missing as we pop them off for easy
+  // traversal back down.
+  std::optional<Dir> dir_tmp;
```

Comment:
```suggestion
  std::optional<Dir> work_dir;
```

("tmp" has a specific meaning that doesn't apply here)